drivers/usb/udc/udc_ambiq.c

```diff
@@ -0,0 +1,960 @@
+	bool ctrl_pending_in_ack;
+	bool ctrl_pending_setup;
+	bool ctrl_setup_recv_at_status_in;
+};
+
+struct udc_ambiq_config {
+	uint32_t num_endpoints;
+	int speed_idx;
+	struct udc_ep_config *ep_cfg_in;
+	struct udc_ep_config *ep_cfg_out;
+	struct gpio_dt_spec usb_vddusb33_gpio;
+	struct gpio_dt_spec usb_vddusb0p9_gpio;
+	void (*make_thread)(const struct device *dev);
+	void (*irq_enable_func)(const struct device *dev);
+	void (*irq_disable_func)(const struct device *dev);
+	void (*callback_register_func)(const struct device *dev);
+};
+
+static int udc_ambiq_rx(const struct device *dev, uint8_t ep, struct net_buf *buf);
+
+static int usbd_ctrl_feed_dout(const struct device *dev, const size_t length)
+{
+	struct udc_ep_config *cfg = udc_get_ep_cfg(dev, USB_CONTROL_EP_OUT);
+	struct net_buf *buf;
+
+	buf = udc_ctrl_alloc(dev, USB_CONTROL_EP_OUT, length);
+	if (buf == NULL) {
+		return -ENOMEM;
+	}
+
+	k_fifo_put(&cfg->fifo, buf);
+	if (length) {
+		udc_ambiq_rx(dev, cfg->addr, buf);
+	}
+
+	return 0;
+}
+
+static int udc_ambiq_tx(const struct device *dev, uint8_t ep, struct net_buf *buf)
+{
+	const struct udc_ambiq_data *priv = udc_get_private(dev);
+	uint32_t status;
+
+	if (udc_ep_is_busy(dev, ep)) {
+		LOG_WRN("ep 0x%02x is busy!", ep);
+		return 0;
+	}
+	udc_ep_set_busy(dev, ep, true);
+
+	/* buf equals NULL is used as indication of ZLP request */
+	if (buf == NULL) {
+		status = am_hal_usb_ep_xfer(priv->usb_handle, ep, NULL, 0);
+	} else {
+		status = am_hal_usb_ep_xfer(priv->usb_handle, ep, buf->data, buf->len);
+	}
+
+	if (status != AM_HAL_STATUS_SUCCESS) {
+		udc_ep_set_busy(dev, ep, false);
+		LOG_ERR("am_hal_usb_ep_xfer write failed(0x%02x), %d", ep, (int)status);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int udc_ambiq_rx(const struct device *dev, uint8_t ep, struct net_buf *buf)
+{
+	struct udc_ambiq_data *priv = udc_get_private(dev);
+	struct udc_ep_config *cfg = udc_get_ep_cfg(dev, USB_CONTROL_EP_OUT);
+	uint32_t status;
+	uint16_t rx_size = buf->size;
+
+	if (udc_ep_is_busy(dev, ep)) {
+		LOG_WRN("ep 0x%02x is busy!", ep);
+		return 0;
+	}
+	udc_ep_set_busy(dev, ep, true);
+
+	/* Make sure that OUT transaction size triggered doesn't exceed EP's MPS */
+	if ((ep != USB_CONTROL_EP_OUT) && (cfg->mps < rx_size)) {
+		rx_size = cfg->mps;
+	}
+
+	status = am_hal_usb_ep_xfer(priv->usb_handle, ep, buf->data, rx_size);
+	if (status != AM_HAL_STATUS_SUCCESS) {
+		udc_ep_set_busy(dev, ep, false);
+		LOG_ERR("am_hal_usb_ep_xfer read(rx) failed(0x%02x), %d", ep, (int)status);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static void udc_ambiq_evt_callback(const struct device *dev, am_hal_usb_dev_event_e dev_state)
+{
+	struct udc_ambiq_data *priv = udc_get_private(dev);
+
+	switch (dev_state) {
+	case AM_HAL_USB_DEV_EVT_BUS_RESET:
+		/* enable usb bus interrupts */
+		am_hal_usb_intr_usb_enable(priv->usb_handle,
+					   USB_CFG2_SOFE_Msk | USB_CFG2_ResumeE_Msk |
+						   USB_CFG2_SuspendE_Msk | USB_CFG2_ResetE_Msk);
```

Comment:
```c
		am_hal_usb_intr_usb_enable(priv->usb_handle,
					   USB_CFG2_SOFE_Msk | USB_CFG2_ResumeE_Msk |
					   USB_CFG2_SuspendE_Msk | USB_CFG2_ResetE_Msk);
```

Reply:
clang-format keeps add a tab there. I have manually reverted the change by clang-format and push. Let's see whether the CI will let us through 😄